feat: implement node duplication feature

gentamura opened the pull request:
## Summary
This PR introduces a node duplication feature. For `FileNode`s, this includes a "deep copy" where associated file entities are also replicated in storage.

## Related Issue
Closes #304

## Changes
- **Core File Copying Logic & API (`giselle-engine`):**
  - Implemented `copyFile(workspaceId, sourceFileId, destinationFileId)` in `core/files/copy-file.ts` using `unstorage`'s `getItemRaw` and `setItemRaw` for storage-agnostic file duplication.
  - Exposed this functionality via a new HTTP API endpoint `/api/giselle/copyFile` in `http/router.ts`.
- **Node Creation, Cloning, and Default Name Logic (`@giselle-sdk/node-utils`):**
  - **Consolidated into `@giselle-sdk/node-utils` package:**
    - `node-factories.ts`: Moved from `packages/workflow-designer/src` to this new utility package. It consolidates `create` and `clone` logic for all node types, aiming to prevent omissions when adding new node types and centralizing node manipulation logic.
      - **`create` Method:** Migrated node creation logic previously in `internal-packages/workflow-designer-ui/src/editor/tool/toolbar/state.tsx` to be invoked via `nodeFactories.create`.
      - **`clone` Method:** Implemented specific cloning logic for each node type within `nodeFactories`.
        - *Note: Common parts of the cloning logic could potentially be further consolidated in the future.*
    - `default-name.ts`: Moved from `packages/workflow-utils/src` to this package. It contains default name generation functions (e.g., `defaultName`, `triggerNodeDefaultName`).
  - *This consolidation into `@giselle-sdk/node-utils` improves modularity and helps resolve circular dependencies.*
- **Connection Duplication (`packages/workflow-designer/src/workflow-designer.ts`):**
  - Implemented connection duplication logic directly within the `WorkflowDesigner` core, as `nodeFactories.clone` focuses solely on node definition duplication.
  - Introduced `ConnectionCloneStrategy` (currently defaulting to and primarily using `inputs-only`) to control how connections are duplicated.
    - *Note: This implementation is functional but might benefit from further refinement for more complex strategies or scenarios.*
- **Text Editor Utilities (`@giselle-sdk/text-editor-utils`):**
  - **Relocated to `text-editor-utils` package:** Text editor extensions and utility functions were moved to the new `@giselle-sdk/text-editor-utils` package to resolve circular dependency issues.
- **UI Implementation for Node Duplication (`internal-packages/workflow-designer-ui`):**
  - Added support for node duplication via keyboard shortcuts (Cmd/Ctrl + D).
  - Added a "Duplicate Node" option to the node context menu.
- **`FileNode` Specific Duplication Logic:**
  - **`workflow-designer-context.tsx` (`WorkflowDesignerProvider`):** The `copyNode` method now orchestrates the deep copy for `FileNode`s. It handles the initial UI update for the duplicated node and then asynchronously calls the `client.copyFile` API for storage-side file replication. The file copy status is updated on the node afterward.
  - **`node-factories.ts` (`fileVariableFactoryImpl.clone` in `@giselle-sdk/node-utils`):** When a `FileNode` is cloned, a temporary `originalFileIdForCopy` property is added to each `FileData` object. This ID is used by the `WorkflowDesignerProvider` to identify the source file for the storage copy operation and is removed after the copy attempt.
  - **`use-duplicate-node.ts`:** Parses the `NodeLike` object (obtained from `Workspace.data.nodes`) into a strict `Node` object using `Node.parse()` before passing it to the `copyNode` context function, ensuring type safety for the duplication process.

## Testing
- **UI Test Instructions:**
  - Select a node and press `Cmd + D` (macOS) or `Ctrl + D` (Windows/Linux).
  - Right-click on a node and select "Duplicate Node" from the context menu.
- **Verification:**
  - Confirm that the node is duplicated in the UI almost immediately.
  - For `FileNode`s, verify in Supabase Storage that the associated files have also been duplicated with new IDs after a short delay (due to asynchronous file copying). Check the `FileData` status on the duplicated node if issues arise.
  - Check that duplicated connections adhere to the `ConnectionCloneStrategy` (currently `inputs-only`).

| base | input connection | output connection |
|---|---|---|
| ![SCR-20250521-mkbq](https://github.com/user-attachments/assets/9484d714-104c-4f79-a516-79850187e126) | ![SCR-20250521-mkfu](https://github.com/user-attachments/assets/1aa77c68-95cd-4d14-a4f5-34705b8f29f6) | ![SCR-20250521-mkiy](https://github.com/user-attachments/assets/5f3aaac1-ff40-4e97-98b9-d5b89dfc66a5) |

## Other Information
N/A

gentamura commented on internal-packages/workflow-designer-ui/src/editor/properties-panel/text-generation-node-properties-panel/input-panel.tsx:
```diff
@@ -6,6 +6,7 @@ import {
 	type TextGenerationNode,
 } from "@giselle-sdk/data-type";
 import { isJsonContent, jsonContentToText } from "@giselle-sdk/text-editor";
+import { defaultName } from "@giselle-sdk/workflow-utils";
```
Since defaultName is used below, this review is not necessary.

https://github.com/giselles-ai/giselle/blob/66c4d5a28627b4652954e8528bf426328fe3dbf0/internal-packages/workflow-designer-ui/src/editor/properties-panel/text-generation-node-properties-panel/input-panel.tsx#L181
https://github.com/giselles-ai/giselle/blob/66c4d5a28627b4652954e8528bf426328fe3dbf0/internal-packages/workflow-designer-ui/src/editor/properties-panel/text-generation-node-properties-panel/input-panel.tsx#L209

gentamura commented on internal-packages/workflow-designer-ui/src/editor/properties-panel/ui/properties-panel.tsx:
```diff
@@ -1,9 +1,9 @@
 "use client";
 
 import type { Node } from "@giselle-sdk/data-type";
+import { defaultName } from "@giselle-sdk/workflow-utils";
```
Since defaultName is used below, this review is not necessary.

https://github.com/giselles-ai/giselle/blob/66c4d5a28627b4652954e8528bf426328fe3dbf0/internal-packages/workflow-designer-ui/src/editor/properties-panel/ui/properties-panel.tsx#L52

toyamarinyon commented on packages/workflow-designer/src/react/workflow-designer-context.tsx:
```diff
@@ -132,6 +146,84 @@ export function WorkflowDesignerProvider({
+					},
+				);
+
+				const resolvedFiles = await Promise.all(fileCopyPromises);
+				const newContentForNode: FileContent = {
+					...newFileNode.content,
+					files: resolvedFiles,
+				};
+
```
Since there is no feedback and nodes are not duplicated until the file copy is complete, which makes it difficult to know if it's working properly, I want to duplicate the nodes (display them in the UI) before copying the files, and then execute the file copy

toyamarinyon commented on packages/workflow-utils/src/default-name.ts:
"I want to limit workflow-utils to responsibilities related only to workflow construction, so I don't want to include defaultName. I think the current location is also inappropriate, but if we were to move it, I think workflow-designer would be more appropriate.

toyamarinyon commented on packages/text-editor/package.json:
```diff
@@ -35,6 +35,7 @@
 	},
 	"dependencies": {
 		"@giselle-sdk/data-type": "workspace:^",
+		"@giselle-sdk/workflow-utils": "workspace:^",
```
"I don't want to create a dependency on workflow-utils just for defaultName. It might be better to create a package like node-utils and place node-factory and defaultName there.

toyamarinyon reviewed (commented):
Thank you for your response. I have confirmed it's working in the Preview environment. I've added comments about the implementation, so please check and address them.

gentamura commented on packages/giselle-engine/src/http/router.ts:
```diff
@@ -144,6 +144,23 @@ export const createJsonRouters = {
+				return new Response(null, { status: 204 });
+			},
+		}),
+	copyFile: (giselleEngine: GiselleEngine) =>
+		createHandler({
+			input: z.object({
+				workspaceId: WorkspaceId.schema,
+				sourceFileId: FileId.schema,
+				destinationFileId: FileId.schema,
+			}),
+			handler: async ({ input }) => {
+				await giselleEngine.copyFile(
+					input.workspaceId,
+					input.sourceFileId,
+					input.destinationFileId,
+				);
+
 				return new Response(null, { status: 204 });
 			},
```
:memo: Since refine is not used in this scope, this comment should be resolve.

gentamura commented on packages/giselle-engine/src/core/files/copy-file.ts:
```diff
@@ -0,0 +1,62 @@
+ * @returns A promise that resolves when the file is copied.
+ * @throws Error if reading the source file or writing the destination file fails.
+ */
+export async function copyFile(args: {
+	context: GiselleEngineContext;
+	workspaceId: WorkspaceId;
+	sourceFileId: FileId;
+	destinationFileId: FileId;
+}) {
```
:memo: Returning an empty return is undesirable; an error should be thrown, but we consider it unnecessary here.

gentamura commented on packages/node-utils/src/node-factories.ts:
```diff
@@ -0,0 +1,528 @@
+				console.error("Error processing prompt for TextGeneration clone:", e);
+				clonedContent.prompt = "";
+			}
+		}
```
Fixed 4fcfa497dd65

gentamura commented on packages/text-editor-utils/src/extensions/index.ts:
Since importing tiptap extensions as external packages caused issues with resolving type information, we replaced them with the Starter Kit. We have selected the same extensions for use as before.

https://tiptap.dev/docs/editor/extensions/functionality/starterkit

gentamura commented on packages/text-editor-utils/src/extensions/index.ts:
```diff
@@ -22,18 +12,16 @@ lowlight.register("js", js);
-	ListItem,
-	BulletList,
-	Code,
+	StarterKit.configure({
+		blockquote: false,
+		codeBlock: false,
+		hardBreak: false,
+		heading: false,
+		horizontalRule: false,
+		dropcursor: false,
+		gapcursor: false,
+	}),
```
Since importing tiptap extensions as external packages caused issues with resolving type information, we replaced them with the Starter Kit. We have selected the same extensions for use as before.

https://tiptap.dev/docs/editor/extensions/functionality/starterkit

gentamura commented:
@toyamarinyon cc: @shige Thanks for the review! I've updated code and the PR description. Regarding the packages, here are the key topics. Please review again:

- Moved `default-name.ts` (from `workflow-utils`) and `node-factories.ts` (from `workflow-designer`) to the new `@giselle-sdk/node-utils` package.
- Relocated text editor extensions and utils to the new `@giselle-sdk/text-editor-utils` package.
  - This separation into `text-editor-utils` was done to resolve circular package dependencies.

gentamura commented:
This PR has been closed and split into separate PRs:

- https://github.com/giselles-ai/giselle/pull/974
- https://github.com/giselles-ai/giselle/pull/977
- https://github.com/giselles-ai/giselle/pull/1003
